test/Test/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,180 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Database.LSMTree.Internal.Chunk (tests) where
+
+import           Prelude hiding (concat, drop, length)
+
+import           Control.Arrow ((>>>))
+import           Control.Monad.ST.Strict (runST)
+import qualified Data.List as List (concat, drop, length)
+import           Data.Maybe (catMaybes, fromJust, isJust, isNothing)
+import           Data.Vector.Primitive (Vector, concat, fromList, length,
+                     toList)
+import           Data.Word (Word8)
+import           Database.LSMTree.Internal.Chunk (Chunk, createBaler, feedBaler,
+                     fromChunk, unsafeEndBaler)
+import           Test.QuickCheck (Arbitrary (arbitrary, shrink),
+                     NonEmptyList (NonEmpty), Positive (Positive, getPositive),
+                     Property, Small (Small, getSmall), Testable, scale,
+                     shrinkMap, tabulate, (==>))
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+-- * Test case generation and shrinking
+
+instance Arbitrary (Vector Word8) where
+
+    arbitrary = fromList <$> arbitrary
+
+    shrink = shrinkMap fromList toList
+
+{-
+    The type of minimum chunk sizes.
+
+    This type is isomorphic to 'Int' but has a different way of generating test
+    cases. Only small, positive integers are generated but with the size
+    parameter squared.
+
+    The reasons for the squaring of the size parameter are somewhat subtle.
+
+    First, we want the ratio between the average minimum chunk size and the
+    average size of a data portion that we feed to a baler to be independent of
+    the size parameter. Each data portion is a list of primitive vectors of
+    bytes, and arbitrarily generated lists and byte vectors have lengths that
+    are small, positive integers. Such integers are \(s/2\) on average, where
+    \(s\) denotes the size parameter. As a result, the average size of data fed
+    to a baler is \(s^{2}/4\). By generating minimum chunk sizes with the size
+    parameter squared, the average minimum chunk size is \(s^{2}/2\) and
+    therefore twice the average size of a data portion fed, independently
+    of \(s\).
```

Comment:
Nice that you documented your reasoning! A few thoughts on that, only take what you think is useful.

Hmm, so the test input size in bytes scales `O(s^3)` with the size parameter `s`. List of list of vector. Do the tests benefit much from individual vectors being long? Maybe we could generate them with size 0 to 10? Then the minimum chunk size can just scale linearly.

Generally the number of inputs for each chunk (basically `buildupPhasesLengths` below) is quite low:
```
      Lengths of buildup phases (1055 in total):
      39.05% 0
      26.82% 1
      16.40% 2
      10.05% 3
       4.74% 4
       2.27% 5
       0.57% 6
       0.09% 7
```

We definitely want to cover the very small numbers (especially 0), but it would be nice to also have larger ones, so I'd be inclined to increase the min chunk size a bit. Maybe also more so the larger the size parameter gets (e.g. scale with `(** 2.5)`). This means that `prop_noRemnantAfterOutput` needs to discard a few more inputs, but that should be okay.

Reply:
> Hmm, so the test input size in bytes scales `O(s^3)` with the size parameter `s`. List of list of vector. Do the tests benefit much from individual vectors being long? Maybe we could generate them with size 0 to 10? Then the minimum chunk size can just scale linearly.

Just fixing the size parameter for some data seems to be questionable to me: it’s somehow not in line with the QuickCheck way. 🙂 That said, with only 100 tests, the size parameter stays fairly small, and so I expect the total size of generated baler input to be at most a few kilobytes.

Reply:
> Generally the number of inputs for each chunk (basically `buildupPhasesLengths` below) is quite low: […]
>
> We definitely want to cover the very small numbers (especially 0), but it would be nice to also have larger ones, so I'd be inclined to increase the min chunk size a bit. Maybe also more so the larger the size parameter gets (e.g. scale with `(** 2.5))`. This means that `prop_noRemnantAfterOutput` needs to discard a few more inputs, but that should be okay.

Hmm, initially I thought that choosing a higher exponent is wrong, but now I realize that it might be in line with the idea of the size parameter: you might want not just the sizes of data structures but also other interesting parameters to grow when the size parameter grows, and one such interesting parameter could be the buildup phase length. If we want it to grow *linearly* with the size parameter, to stick with the usual QuickCheck approach, we should pick 3 as the exponent. I’ll perform a few experiments to see how things change with different derivations  of the size parameter for minimum chunk size generation.

Reply:
Meanwhile, I experimented with different formulas for deriving the size parameter for minimum chunk size generation.

Generally, I found it hard to see something like a clear picture in the results. At least, I somewhat got the impression that raising the exponent doesn’t have clear benefits: it seems to help only little for `prop_noRemnantAfterOutput` while bearing the risk of having too strong an impact for `prop_remnantChunkIsNonEmpty` and `prop_remnantChunkIsSmall`. Using a constant factor greater than 1 seems to have similar effects, although the details are apparently a bit different.

In the end, I came to the conclusion that the best way forward might be to keep things simple while alleviating the problem with the test case distribution at least slightly: just stick with squaring but furthermore multiply the result with 2. This way, the test case distributions aren’t too bad, and we don’t dive into more complicated approaches that we don’t really understand.

Reply:
I’ve implemented the above-mentioned solution. The result is in f3229bee1a9a3bedc6b8bcbb19860eecc7fddcd7.